universe/Universe.cpp

```diff
@@ -1660,6 +1660,17 @@ void Universe::ForgetKnownObject(int empire_id, int object_id) {
     objects.Remove(object_id);
 }
 
+void Universe::GrantShipDesignKnowledge(int empire_id, int object_id) {
+    std::shared_ptr<const Ship> ship = GetShip(object_id);
+    if (!ship) return;
+    int design_id = ship->DesignID();
+    if (design_id == ShipDesign::INVALID_DESIGN_ID) {
+        ErrorLogger() << "GrantShipDesignKnowledge got invalid design id for ship with id " << object_id;
+    } else {
+        m_empire_known_ship_design_ids[empire_id].insert(design_id);
```

Comment:
Pondering if it's worth calling `Universe::SetEmpireKnowledgeOfShipDesign` instead of re-implementing some of the checks here... You'd also get some extra detailed error messages instead of silently returning if the ship doesn't exist, and would get a check that `empire_id` isn't `ALL_EMPIRES`, which probably should be checked for.

Also, the name is oddly inconsistent with that function's naming... Maybe instead it could be `SetEmpireKnowledgeOfDesignOfShip` ?

And perhaps change the parameter name to `ship_id` instead of `object_id` ?